mongo_connector/dest_mapping.py

```diff
@@ -0,0 +1,169 @@
+import logging
+import threading
+import re
+from mongo_connector import errors
+
+
+LOG = logging.getLogger(__name__)
+
+"""New structure to handle both plain and
+wildcard mapping namespaces dynamically.
+"""
+
+
+class DestMapping():
+    def __init__(self, namespace_set, ex_namespace_set, user_mapping):
+        """
+            namespace_set and ex_namespace_set will not be non-empty
+            in the same time.
+            user_mapping should be non-empty together with namespace_set.
+            This assumption has been verified in the get_config_options()
+            in connector.py.
+        """
+        # a dict containing plain mappings
+        self.plain = {}
+        # a dict containing wildcard mappings
+        self.wildcard = {}
+        # a dict containing reverse plain mapping
+        # because the mappings are not duplicated,
+        # so the values should also be unique
+        self.reverse_plain = {}
+        # a dict containing plain db mappings, db -> a set of mapped db
+        self.plain_db = {}
+
+        # the input namespace_set and ex_namespace_set could contain wildcard
+        self.namespace_set = namespace_set
+        self.ex_namespace_set = ex_namespace_set
+
+        self.lock = threading.Lock()
+
+        # initialize
+        for ns in namespace_set:
+            user_mapping.setdefault(ns, ns)
+
+        for k, v in user_mapping.items():
+            self.set(k, v)
+
+    def set_plain(self, key, value):
+        """A utility function to set the corresponding plain variables"""
+        db, col = key.split(".", 1)
+        self.plain[key] = value
+        self.reverse_plain[value] = key
+        if col != "$cmd":
+            if db not in self.plain_db:
+                self.plain_db[db] = set([value.split(".")[0]])
+            else:
+                self.plain_db[db].add(value.split(".")[0])
+
+    def match(self, src, dst_pattern):
+        """If source string src matches dst, return the matchobject"""
+        reg_pattern = r'\A' + dst_pattern.replace('*', '(.*)') + r'\Z'
+        m = re.match(reg_pattern, src)
+        return m
+
+    def match_set(self, plain_src, dst_arr):
+        for x in dst_arr:
+            if plain_src == x:
+                return True
+            if "*" in x:
+                m = self.match(plain_src, x)
+                if m:
+                    return True
+
+        return False
+
+    def replace(self, src_match, map_pattern):
+        """Given the matchobject src_match,
+        replace corresponding '*' in map_pattern with src_match."""
+        wildcard_matched = src_match.group(1)
+        return map_pattern.replace("*", wildcard_matched)
+
+    def get(self, plain_src_ns, defaultval=""):
+        """Given a plain source namespace, return a mapped namespace if matched.
+        If no match and given defaultval, return it as a default value.
+        """
+        self.lock.acquire()
```

Comment:
Acquire the lock using a with statement to simplify this code. See, https://docs.python.org/2/library/threading.html#using-locks-conditions-and-semaphores-in-the-with-statement